src/Castle.Facilities.AspNet.Mvc/AspNetMvcControllerFactory.cs

```diff
@@ -0,0 +1,93 @@
+// Copyright 2004-2017 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Reflection;
+
+namespace Castle.Facilities.AspNet.Mvc
+{
+	using System;
+	using System.Web.Mvc;
+	using System.Web.Routing;
+	using System.Web.SessionState;
+
+	using Castle.Facilities.AspNet.Mvc.Exceptions;
+	using Castle.MicroKernel;
+
+	public class AspNetMvcControllerFactory : IControllerFactory
+	{
+		private readonly IKernel container;
+		private Assembly controllerAssembly;
+
+		public event Action<IKernel, Type> BeforeControllerResolved;
+		public event Action<IKernel, object> AfterControllerReleased;
```

Comment:
Not really a fan of all these events firing back and forth between the controller factory and the facility, feels like something is in the wrong place. Maybe the controller factory should create and destroy the scope.